csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_pad.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+struct Edge {
+  Expr* expr_ = nullptr;
+  size_t index_ = 0;
+
+  Edge(Expr* expr, size_t index) : expr_(expr), index_(index) {}
+
+  Val* val() const {
+    return expr_->input(index_);
+  }
+};
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `unaryOp(0) == 0`
+bool padCompatibleUnaryOp(UnaryOpType t) {
+  switch (t) {
+    case UnaryOpType::Cast:
```

Comment:
Do we handle this using `SimplifyingIrBuilder::maybeCastExpr(pad_value)` somewhere?

Reply:
Would you mind elaborate on this one?

For the context, I'm not trying to simplify existing `cast` in the IR. But merely move the `PadOp` from the cast output to its input.

Reply:
Good point.

I didn't know that `pad_value` is sensitive with its DataType. I'll look into that.